tests/util/bft.py

```diff
@@ -393,6 +393,46 @@ def force_quorum_including_replica(self, replica_id, primary=0):
                     nursery.start_soon(self._assert_state_transfer_not_started,
                                        r)
 
+    async def assert_fast_path_prevalent(self, as_of_seq_num):
+        metric_key = ['replica', 'Counters', 'slowPathCount']
+        nb_slow_paths = await self.metrics.get(0, *metric_key)
+
+        assert nb_slow_paths - as_of_seq_num <= 1, \
+            f'Fast path is not prevalent for n={self.config.n}, f={self.config.f}, c={self.config.c}.'
+
+    async def assert_slow_path_prevalent(self, as_of_seq_num):
+        metric_key = ['replica', 'Gauges', 'lastExecutedSeqNum']
+        all_executed_sequences = await self.metrics.get(0, *metric_key)
+
+        metric_key = ['replica', 'Counters', 'slowPathCount']
+        nb_slow_paths = await self.metrics.get(0, *metric_key)
+
+        assert nb_slow_paths >= all_executed_sequences - as_of_seq_num, \
+            f'Slow path is not prevalent for n={self.config.n}, f={self.config.f}, c={self.config.c}.'
+
+    async def write_known_kv(self, protocol):
```

Comment:
Passing the protocol and using it in the BFT tester breaks abstractions. The reasoning is that the protocol is tied specifically to the skvbc implementation, while the rest of the test network code is supposed to be usable with any state machine. The protocol itself is a concrete type and not an abstract interface here, as it's specifically tied to a state machine that allows transactions and reading of blocks.

This isn't the first time abstraction has been broken in this module (see `assert_successful_put_get` and `prime_for_state_transfer`) I'm fine leaving this here for this PR and fixing this in a future one.

However, the intention was that this module be generic way of running code and performing tests that are bft specific but agnostic to the state machine. We should have another module for specific functions like this that is part of skvbc. For places in this code where we need specific functionality such as "performing a write", we should either move that to a higher level module or create an [abstract type](https://docs.python.org/3/library/abc.html) when we truly want to share the functionality across state machines.

Reply:
Agreed, I also thought that bft.py is in charge of more than it should be. I'll track this in a separate tech debt task.